File: tools/operator-config.mjs
import { existsSync, readFileSync } from 'node:fs';
import path from 'node:path';

const DISABLED = new Set(['0', 'off', 'none', 'false', 'disabled']);

function usable(value) {
  const text = String(value ?? '').trim();
  if (!text || DISABLED.has(text.toLowerCase())) return '';
  if (text.startsWith('_')) return '';
  return text.replace(/\/+$/, '');
}

function hubField(file, key, envKeys = []) {
  const hub = file && typeof file.hub === 'object' && file.hub ? file.hub : {};
  for (const envKey of envKeys) {
    const fromEnv = usable(process.env[envKey]);
    if (fromEnv) return fromEnv;
  }
  return usable(hub[key]) || usable(file[key]);
}

export function loadOperatorConfig(root) {
  const filePath = path.join(root, '_local', 'operator.json');
  let file = {};
  if (existsSync(filePath)) {
    try {
      file = JSON.parse(readFileSync(filePath, 'utf8'));
    } catch {
      file = {};
    }
  }
  const nas = file && typeof file.nas === 'object' && file.nas ? file.nas : {};
  const deploymentPhase = usable(file.deployment_phase) || 'operator_pc';
  return {
    deployment_phase: deploymentPhase,
    openclaw_adapter_base_url: hubField(file, 'openclaw_adapter_base_url', [
      'MY_AGENT_OPENCLAW_ADAPTER_BASE_URL',
      'OPENCLAW_ADAPTER_BASE_URL',
    ]),
    brand_manual_url: hubField(file, 'brand_manual_url', ['MY_AGENT_BRAND_MANUAL_URL']),
    product_data_base_url: hubField(file, 'product_data_base_url', [
      'MY_AGENT_PRODUCT_DATA_BASE_URL',
    ]),
    adapter_auth: {
      mode: usable(file?.adapter_auth?.mode) || 'install_bootstrap',
      bootstrap_path: usable(file?.adapter_auth?.bootstrap_path) || '/cqr/adapter/auth/bootstrap',
      bootstrap_key: usable(process.env.MY_AGENT_ADAPTER_BOOTSTRAP_KEY)
        || usable(file?.adapter_auth?.bootstrap_key),
    },
    adapter_progress: {
      request_path: usable(file?.adapter_progress?.request_path) || '/cqr/adapter/request',
      status_path_template: usable(file?.adapter_progress?.status_path_template)
        || '/cqr/adapter/jobs/{job_id}',
      poll_interval_ms: Number(file?.adapter_progress?.poll_interval_ms) || 2500,
      timeout_ms: Number(file?.adapter_progress?.timeout_ms) || 1800000,
    },
    nas,
  };
}

/** URLs baked into published org module (deploy-overrides + module.json). */
export function operatorHubForPublish(root) {
  const op = loadOperatorConfig(root);
  return {
    deployment_phase: op.deployment_phase,
    openclaw_adapter_base_url: op.openclaw_adapter_base_url,
    brand_manual_url: op.brand_manual_url,
    product_data_base_url: op.product_data_base_url,
    adapter_auth: op.adapter_auth,
    adapter_progress: op.adapter_progress,
  };
}

File: tools/operator_config.py
"""Operator/environment settings. This file has no machine-specific hosts."""

from __future__ import annotations

import json
import os
from pathlib import Path

DISABLED = frozenset({"0", "off", "none", "false", "disabled"})


def find_repo_root(start: Path | None = None) -> Path:
    start = (start or Path.cwd()).resolve()
    for path in [start, *start.parents]:
        if (path / "manifest.json").is_file() and (path / "agent-module").is_dir():
            return path
    here = Path(__file__).resolve().parent.parent
    if (here / "manifest.json").is_file() and (here / "agent-module").is_dir():
        return here
    raise FileNotFoundError("organization module repo root not found")


def load_operator_config(root: Path | None = None) -> dict:
    root = root or find_repo_root()
    path = root / "_local" / "operator.json"
    if not path.is_file():
        return {}
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return {}
    return data if isinstance(data, dict) else {}


def first_nonempty(*values: object) -> str:
    for value in values:
        text = str(value or "").strip()
        if text and text.lower() not in DISABLED:
            return text
    return ""


def _nas(root: Path | None = None) -> dict:
    cfg = load_operator_config(root)
    nas = cfg.get("nas")
    return nas if isinstance(nas, dict) else {}


def brand_manual_url(root: Path | None = None) -> str:
    cfg = load_operator_config(root)
    hub = cfg.get("hub") if isinstance(cfg.get("hub"), dict) else {}
    return first_nonempty(
        os.environ.get("MY_AGENT_BRAND_MANUAL_URL"),
        hub.get("brand_manual_url") if isinstance(hub, dict) else None,
        cfg.get("brand_manual_url"),
    )


def product_data_base_url(root: Path | None = None) -> str:
    cfg = load_operator_config(root)
    hub = cfg.get("hub") if isinstance(cfg.get("hub"), dict) else {}
    return first_nonempty(
        os.environ.get("MY_AGENT_PRODUCT_DATA_BASE_URL"),
        hub.get("product_data_base_url") if isinstance(hub, dict) else None,
        cfg.get("product_data_base_url"),
    )


def openclaw_adapter_base_url(root: Path | None = None) -> str:
    cfg = load_operator_config(root)
    hub = cfg.get("hub") if isinstance(cfg.get("hub"), dict) else {}
    return first_nonempty(
        os.environ.get("MY_AGENT_OPENCLAW_ADAPTER_BASE_URL"),
        os.environ.get("OPENCLAW_ADAPTER_BASE_URL"),
        hub.get("openclaw_adapter_base_url") if isinstance(hub, dict) else None,
        cfg.get("openclaw_adapter_base_url"),
    )


def deployment_phase(root: Path | None = None) -> str:
    cfg = load_operator_config(root)
    return first_nonempty(cfg.get("deployment_phase")) or "operator_pc"


def nas_root(root: Path | None = None) -> Path | None:
    text = first_nonempty(os.environ.get("CQR_NAS_ROOT"), _nas(root).get("root"))
    return Path(text) if text else None


def nas_alllisting(root: Path | None = None) -> Path | None:
    nas = _nas(root)
    text = first_nonempty(os.environ.get("CQR_NAS_ALLLISTING"), nas.get("alllisting"))
    return Path(text) if text else None


def nas_dev_xlsx(root: Path | None = None) -> Path | None:
    nas = _nas(root)
    text = first_nonempty(os.environ.get("CQR_DEV_XLSX"), nas.get("dev_xlsx"))
    return Path(text) if text else None


def nas_po_dir(root: Path | None = None) -> Path | None:
    nas = _nas(root)
    text = first_nonempty(os.environ.get("CQR_PO_DIR"), nas.get("po_clothing"))
    return Path(text) if text else None


def listing_file(root: Path | None = None) -> Path | None:
    text = first_nonempty(os.environ.get("CQR_LISTING"))
    if text:
        path = Path(text)
        return path if path.is_file() else None
    return None


def resolve_listing_file(root: Path | None = None) -> Path | None:
    root = root or find_repo_root()
    explicit = listing_file(root)
    if explicit:
        return explicit
    candidates: list[Path] = []
    cache = root / "_local" / "cache" / "listings"
    if cache.is_dir():
        candidates.extend(cache.glob("*alllisting*.txt"))
        candidates.extend(cache.glob("*.txt"))
    folder = nas_alllisting(root)
    if folder is not None:
        try:
            if folder.exists():
                candidates.extend(folder.glob("*alllisting*.txt"))
                candidates.extend(folder.glob("*.txt"))
        except OSError:
            pass
    files = [path for path in candidates if path.is_file()]
    if not files:
        return None
    files.sort(key=lambda p: p.stat().st_mtime, reverse=True)
    return files[0]
